prov/gni/src/gnix_vector.c

```diff
@@ -0,0 +1,420 @@
+		return -FI_ENOMEM;
+	} else {
+		attr->cur_size = attr->vec_initial_size;
+	}
+
+	memcpy(&vec->attr, attr, sizeof(gnix_vec_attr_t));
+
+	return FI_SUCCESS;
+}
+
+static inline void __gnix_vec_close_entries(gnix_vector_t *vec)
+{
+	int i;
+
+	for (i = 0; i < vec->attr.cur_size; i++) {
+		if (vec->vector[i])
+			vec->attr.destructor(vec->vector[i]);
+		vec->vector[i] = NULL;
+	}
+}
+
+static inline void __gnix_vec_close(gnix_vector_t *vec)
+{
+	free(vec->vector);
+	vec->ops = NULL;
+	vec->attr.cur_size = 0;
+	vec->state = GNIX_VEC_STATE_DEAD;
+}
+
+/*******************************************************************************
+ * LOCKLESS FNS
+ ******************************************************************************/
+static int __gnix_vec_lf_init(gnix_vector_t *vec, gnix_vec_attr_t *attr)
+{
+	int ret;
+
+	ret = __gnix_vec_create(vec, attr);
+	vec->ops = &__gnix_vec_lockless_ops;
+	vec->state = GNIX_VEC_STATE_READY;
+
+	return ret;
+}
+
+static int __gnix_vec_lf_close(gnix_vector_t *vec)
+{
+	__gnix_vec_close_entries(vec);
+	__gnix_vec_close(vec);
+
+	return FI_SUCCESS;
+}
+
+static int __gnix_vec_lf_resize(gnix_vector_t *vec, uint64_t size)
+{
+	return __gnix_vec_resize(vec, size);
+}
+
+static int __gnix_vec_lf_insert_first(gnix_vector_t *vec,
+				      gnix_vec_entry_t *entry)
+{
+	return _gnix_vec_insert_first(vec, entry);
+}
+
+static int __gnix_vec_lf_insert_last(gnix_vector_t *vec,
+			      gnix_vec_entry_t *entry)
+{
+	return _gnix_vec_insert_last(vec, entry);
+}
+
+static int __gnix_vec_lf_insert_at(gnix_vector_t *vec,
+				   gnix_vec_entry_t *entry,
+				   gnix_vec_index_t index)
+{
+	return _gnix_vec_insert_at(vec, entry, index);
+}
+
+static int __gnix_vec_lf_remove_first(gnix_vector_t *vec)
+{
+	return _gnix_vec_remove_first(vec);
+}
+
+static int __gnix_vec_lf_remove_last(gnix_vector_t *vec)
+{
+	return _gnix_vec_remove_last(vec);
+}
+
+static int __gnix_vec_lf_remove_at(gnix_vector_t *vec,
+				   gnix_vec_index_t index)
+{
+	return _gnix_vec_remove_at(vec, index);
+}
+
+static int __gnix_vec_lf_first(gnix_vector_t *vec, void **element)
+{
+	return _gnix_vec_first(vec, element);
+}
+
+static int __gnix_vec_lf_last(gnix_vector_t *vec, void **element)
+{
+	return _gnix_vec_last(vec, element);
+}
+
+static int __gnix_vec_lf_at(gnix_vector_t *vec, void **element, gnix_vec_index_t index)
+{
+	return _gnix_vec_at(vec, element, index);
+}
+
+/*******************************************************************************
+ * LOCKED FNS
+ ******************************************************************************/
+static int __gnix_vec_lk_init(gnix_vector_t *vec, gnix_vec_attr_t *attr)
+{
+	int ret;
+
+	rwlock_init(&vec->lock);
+	rwlock_wrlock(&vec->lock);
+
+	ret = __gnix_vec_create(vec, attr);
+	vec->ops = &__gnix_vec_locked_ops;
+	vec->state = GNIX_VEC_STATE_READY;
+
+	rwlock_unlock(&vec->lock);
+
+	return ret;
+}
+
+static int __gnix_vec_lk_close(gnix_vector_t *vec)
+{
+	rwlock_wrlock(&vec->lock);
```

Comment:
Yes, I'm not sure how else to do this.  If the user closes the vector while they're still using it then the waiting vector calls will return -FI_EINVAL after the lock is released in `__gnix_vec_lk_close` since the vector will be in a DEAD state.

On Fri, 04 Mar 2016 08:23:50 -0800
James Swaro <notifications@github.com> wrote:

> > +
> > +	rwlock_init(&vec->lock);
> > +	rwlock_wrlock(&vec->lock);
> > +
> > +	ret = __gnix_vec_create(vec, attr);
> > +	vec->ops = &__gnix_vec_locked_ops;
> > +	vec->state = GNIX_VEC_STATE_READY;
> > +
> > +	rwlock_unlock(&vec->lock);
> > +
> > +	return ret;
> > +}
> > +
> > +static int __gnix_vec_lk_close(gnix_vector_t *vec)
> > +{
> > +	rwlock_wrlock(&vec->lock);
> 
> close should be done while under exclusive modification, true, but in this circumstance, anyone waiting on the lock will be operating on a dead object after the the lock is acquired. Is this what you intended?
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/ofi-cray/libfabric-cray/pull/696/files#r55051052


-- 
Evan Harvey